### PR TITLE
Log view: wrap lines and improve UX

### LINKED ISSUE #1285: log: enable word wrapping
Hey, 

Would it be possible to enable line wrapping in the `job*` file output? For some reason there are often extremely long lines in output from the models I run (see screenshot). Just wondering. 

Cheers,

Jonny

image

### LINKED ISSUE #1667: log: log view contents is duplicated on reconnect
A bug spotted whilst working on a flaky connection.

Whenever the connection is lost, the UI will attempt to reconnect, when it succeeds all subscriptions are automatically re-submitted.

Unfortunately, the log view does not wipe its local data store when this occurs so when the subscription is resumed, the log file lines returned are appended to the end of the file. Apparently this can also be replicated in offline mode.

It's worth checking whether the analysis view, which also uses a local store, can fall victim to this. It would be good to establish a pattern to make it easier for such views to reset their stores on disconnect.

### PR DESCRIPTION
**Pull request by MetRonnie**
Addresses numerous usability concerns raised by users:

- Close #1285 
- Close #1667
- Made the log display area scroll only, leaving the toolbar pinned to the top
- Moved the file list refresh button next to the file list dropdown

#### Before

image

#### After

image

With word wrapping turned on:

image

#### Check List

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Tests are included
- [x] `CHANGES.md` entry included if this is a change that can affect users

### CODE REVIEW

**Comment by oliver-sanders** [comment]
@MetRonnie, the plan was to use an in-browser text editor / code viewer to provide this functionality along with other things like syntax highlighting, line numbers, etc to avoid implementing each feature individually ourselves.

The umbrella issue for this #144

File: src/components/cylc/log/Log.vue [hunk 1]
```diff
@@ -16,15 +16,18 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 
 <template>
-  <div>
-    <pre><span v-for="(log, index) in computedLogs" :key="index">{{log}}</span></pre>
-  </div>
+  <pre><span
+    v-for="(log, index) in computedLogs"
+    :key="index"
+    :class="wordWrap ? 'text-pre-wrap' : 'text-pre'"
+  >{{ log }}</span></pre>
```
Comment (MetRonnie) [inline src/components/cylc/log/Log.vue:23]: @oliver-sanders This is the only change to the code renderer, applying the `text-pre-wrap` or `text-pre` class

Comment (oliver-sanders) [reply src/components/cylc/log/Log.vue:23]: (not suggesting we don't accept this, but saying we shouldn't go down this road unless it's the only option)

Comment (MetRonnie) [reply src/components/cylc/log/Log.vue:23]: Yeah don't worry, I didn't have any intention of doing so 😄

File: cypress/component/viewToolbar.cy.js [hunk 2]
```diff
@@ -75,7 +75,7 @@ describe('View Toolbar Component', () => {
     // test action=toggle
     cy
       // the toggle should be blue because it's active (default true)
-      .get('.control.toggle .v-btn')
+      .get('[data-cy=control-toggle] .v-btn')
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cypress/component/viewToolbar.cy.js:78]: I argue that what was wrong with this is that there is the potential for unexpected styles being applied as class names are dynamically generated from the `key` field in the control object. When a dev is adding a control they are not likely to be aware that this was happening.

And seeing as this class was not being used for styling but was being used for E2E tests,  I am following Cypress recommended best practice here: https://docs.cypress.io/guides/references/best-practices#Selecting-Elements

> Targeting the element above by tag, class or id is very volatile and highly subject to change. You may swap out the element, you may refactor CSS and update ID's, or you may add or remove classes that affect the style of the element.
>
> Instead, adding the data-cy attribute to the element gives us a targeted selector that's only used for testing.
>
> The data-cy attribute will not change from CSS style or JS behavioral changes, meaning it's not coupled to the behavior or styling of an element.
>
> Additionally, it makes it clear to everyone that this element is used directly by test code.

File: src/views/Log.vue [hunk 3]
```diff
@@ -326,6 +328,18 @@ export default {
...
     const wordWrap = useInitialOptions('wordWrap', { props, emit }, false)
 
+    /** The log subscription results */
+    const results = ref(new Results())
+
+    function reset () {
+      results.value = new Results()
+    }
+
+    whenever(
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply src/views/Log.vue:366]: > I'm guessing `whenever` is just a watcher by a different name?

Correct, so it doesn't work by polling; it is reactive. I have confirmed this by attaching a breakpoint inside the "source" arrow function. I don't know how this actually works, must be the same black magic as computed properties.

**Review by oliver-sanders — APPROVED** [review]

File: cypress/component/viewToolbar.cy.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cypress/component/viewToolbar.cy.js:78]: I'm not arguing that classes should not be used for styling, just that classes should not be automatically added from some other data source. If it was namespaced e.g. ``:class="`control-${key}`"`` that would be better. But seeing as the only use of this class was for testing  I think `data-cy` is more appropriate.

File: src/views/Log.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (MetRonnie) [reply src/views/Log.vue:366]: IMO more concise than
```js
watch(
  () => store.state.offline,
  (val) => {
    if (val) {
      results.value.connected = false
    }
  }
)
```

**Review by markgrahamdawson — APPROVED** [review]

**Pull request merged**